### PR TITLE
Add endpoint to write multiple docs

### PR DESCRIPTION
**Pull request by EricBorczuk**
Add an endpoint to allow writing multiple documents in a single request.

Data sent to this endpoint is expected to be in JSON lines format (1 document per line).

Additionally, you can supply an `id-path` query parameter to use the value at a particular path in each document as the document's key in the database, so if all your documents have an `id` field, you could set `id-path=id` and treat the value in `id` as the document's key. You can also use any valid path syntax (globs not allowed), e.g. `id-path=user.emails.[0].id`

If `id-path` is excluded, random UUID's will be assigned to every document, and the response will have the ID's created corresponding in the same order as the documents were supplied in.

### CODE REVIEW

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:213]: can we add a `@NonNull` now when we have everything validated..

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:411]: can we extract this to some private method, isn't there some other places that uses the same as well? what about inserting the single document?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:433]: so everything related to the `idPath` conversion to the doc path can be moved out of the loop, this way you are not doing the same thing over and over again as this seems to be a constant..

File: testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java
Comment (ivansenic) [inline testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java:699]: can we also add one test that targets the overwrite?

**Review by ivansenic — CHANGES_REQUESTED** [review]
Few small fixes needed imo.. :+1:

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:442]: Why do we require one doc per line? Is it not possible to parse JSON input in a stream fashion?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:444]: The generated ID is ignored if we have an `idPath`, but generating it takes time and drains randomness pools.

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:440]: Do I read it correctly that `CHUNK_SIZE` corresponds to the number of docs in each CQL batch? Why would be break input into multiple batches?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:493]: Could you add a profiling unit test for this, please?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:442]: Technically, the input isn't JSON, since it's actually multiple delimited JSON objects (JSON lines format). I'll look into possible methods to stream JSON lines, but Jackson doesn't have anything for JSON lines afaik

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:440]: I really don't have a reason for this; my only question being: if there are some large number of documents that someone sends, is it fine to put all of those documents in a single batch? is there a point where that would break?

If there isn't really a feasible breaking point, then I can just remove this

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:442]: Another option would be to make the input valid JSON (i.e. require that it's a JSON array) and then streaming probably becomes possible/easy

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:440]: Storage may have limits on max batch size, but I do not think we have visibility into that at SG level.

Since we do not really limit HTTP payload sizes for single doc inserts (do we?) it does not make batch operations worse in terms of overloading the server. It may be worth adding guardrails at HTTP level, but that's a separate concern for sure.

I believe it would not more intuitive for the end user if we had one batch per request.

If it fails at storage level, the user will have to consider using smaller batches, but at least it will be consistent with other APIs for that matter. WDYT?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:442]: Yeah, I guess it may depend on how the library works with input streams. Please post your findings :)

Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:442]: Requiring input to be a JSON array is not unreasonable from my POV.

Comment (gconaty) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:442]: Perhaps an option to import and/or export in either format?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:496]: We do not wait for the `CompletableFuture` associated with the batch operation (or did I miss that), so this catch will not reflect storage-level failure. Would it make sense to wait for batch acknowledgement so that a successful HTTP response would mean the batch was accepted (we should probably use the `202` response code in that case)? WDYT?

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:238]: WDYT about `202` (Accepted) instead? Currently our code does not guarantee that anything was created at this point or will eventually be created (see my other comment about exceptions).

Even if we wait for the ack from storage, batches may or may not be applied by the time we get the ack, so `201` (Created) might be an overstatement. WDYT?

Comment (dougwettlaufer) [reply restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:238]: +1 for `202` since we're just "accepting" here

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:440]: I think that's reasonable, yea - especially the point you made about how stargate doesn't have visibility into any batch size limits in storage.  Also, it's true that the single document write doesn't have a strict limit. So I will remove this!

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (dougwettlaufer) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:184]: > this operation is idempotent

As in a NOOP or it's an upsert?

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (dougwettlaufer) [reply restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:215]: What's the error response from dropwizard look like for that annotation?

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:184]: It will be an upsert - if you provide an `idPath` and do two back-to-back requests, the second request will update instead of inserting

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:496]: It seems like `deleteManyThenInsertBatch` calls `executeBatch`, which then calls

``` 
if (useLoggedBatches) {
      dataStore.batch(queries, ConsistencyLevel.LOCAL_QUORUM).join();
} else {
      dataStore.unloggedBatch(queries, ConsistencyLevel.LOCAL_QUORUM).join();
}
```

This would mean that it is already blocking on some sort of ack, right? If the ack may not mean that data was written, then I agree with changing this to 202...but most probably that should be changed in the other write endpoints too, which I would consider out of scope for this PR; so I think I will leave as 201 just for now, to remain consistent with the other document write endpoints

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java]: So using a JsonParser (Jackson's impl for streaming JSON) is feasible, but only with valid JSON, so I am going to stick with expecting array data over the wire; then, each object in the array can get parsed one by one using an ObjectMapper, which won't pull in much into memory at once :+1: impl is [here](https://github.com/stargate/stargate/pull/1043/files#diff-caf69767609da0fee30d28b2b410ba5e400ed01e1f42af082f495ef82fec8416R435)

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java]: removed!

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java]: added [here](https://github.com/stargate/stargate/pull/1043/files#diff-93b9952f26f7639ef23892552f8d4059d445215cdf6bef3ed4eaacc93357a399R865)

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java]: Ah, right, I missed that :facepalm:

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:462]: I thought we removed "chunks" :thinking: is this a leftover comment?

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (dimas-b) [reply restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:239]: Ok, let's switch to `202` :)

**Review by dimas-b — CHANGES_REQUESTED** [review]
LGTM overall :+1: Just some minor, but critical tweaks remain to be done, I think.

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (dougwettlaufer) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:188]:
```suggestion
        @ApiResponse(code = 202, message = "Accepted", response = MultiDocsResponse.class),
```

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (dougwettlaufer) [inline restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:185]:
```suggestion
      code = 202)
```

**Comment by dougwettlaufer** [comment]
@EricBorczuk would you mind writing up a short example and filing an issue in stargate/docs so we don't forget to document this?

**Review by dougwettlaufer — APPROVED** [review]

**Comment by EricBorczuk** [comment]
Done https://github.com/stargate/docs/issues/132

**Review by dimas-b — APPROVED** [review]

File: restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java
Comment (ivansenic) [reply restapi/src/main/java/io/stargate/web/docsapi/resources/DocumentResourceV2.java:213]: it's generated in `ViolationExceptionMapper`, but it requires a message, please do:

`@NotNull(message = "payload must not be null")`

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (ivansenic) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:457]: shouldn't this be `docsPath.isPresent()`?

**Review by ivansenic — APPROVED** [review]
I think my comments are resolved now..

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:457]: technically it is logically equivalent, but yea that would make way more sense to read :) I'll change

Comment (clun) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:457]: Wondering if we could hit a `JsonToken.END_ARRAY` inside a document itself and loosing track witch the streaming parser here?

```json
[
 { "id":1, "sample":[1,2,3] },
 { "id":2, "sample":[4,5,6] }
]
```

Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java]: I've tested this out, and the parser actually works properly! This is because mapper.readTree(...) on the parser reads the next full object.

Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java]: See https://github.com/stargate/stargate/pull/1043/files#diff-585290b1d8a4cd9f6b594dd2e6a35102c4ab7c86988b9df67452177d38c73868R1, which is used in a few tests

**Pull request merged**
